Querying based on time

tomyyyD opened the pull request:
Range and time queries. A bit large on this PR, but a lot is similar logic that gets the files, so it shouldn't be too hard to look over.

tomyyyD committed:
structure for creating map files

tomyyyD committed:
testing information file

tomyyyD committed:
creating info files and some map stuff:

tomyyyD committed:
map files

tomyyyD committed:
actually adding maps

tomyyyD committed:
Basic Querying

given a time input can use maps to get the file in which we should look for that data

tomyyyD committed:
Fixes to the logging

fixed bug where first folder would have 1 more file than subsequent folders. Added a few test cases for location_from_time

tomyyyD committed:
Fixes and more location_from_time tests

tomyyyD committed:
range queryies and updated time query

tomyyyD committed:
flake8 updates to rexDB and test cases

tomyyyD committed:
cleaning up test cases

tomyyyD committed:
more repo cleanup

tomyyyD committed:
Bug fixes for range querying

also added some more test cases. Fixed the bug that range queries would never find entries in the last file.

tomyyyD committed:
switch default time function to be gmtime

was before localtime which might not return epoch time because it is dependent on time zones. gmtime() will always be time since epoch when converted to a float.

thetazero commented on rexdb.py:
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
+                contents = fd.read()
+                if len(contents) != 0:
+                    lines = int(len(contents) / 12)
+                    for i in range(lines):
```
should be able to read 12 bytes as a time to avoid this indexing arithmetic.

thetazero commented on rexdb.py:
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
+        found_file = True
+        for folder in folders:
+            try:
+                with open(f"db_{self.db_num}/{folder}/{folder}.map", "rb") as fd:
```
Maybe try `f"db_{self.db_num}/{folder}/.map"`

thetazero commented on rexdb.py:
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
+                        lines = int(len(contents) / 12)
+                        for i in range(lines):
+                            (start_time, end_time, num) = struct.unpack("iii", contents[i * 12: i * 12 + 12])
+                            if (start <= start_time and start_time <= end or start <= end_time and end_time <= end):
```
add parentheses because I have skill issues and don't remember  order of operations for and and or.

thetazero replied:
Can also do the python thing of
```python
(start_time <= start <= end_time) or (start_time <= end <= end_time)
```

thetazero committed:
Time faking to speed up unit tests

thetazero committed:
Lint

thetazero commented on rexdb.py:
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
+                contents = fd.read()
+                if len(contents) != 0:
+                    lines = int(len(contents) / 12)
+                    for i in range(lines):
```
I mean something like this:
```python3
while (data := f.read(12)) and len(data) == 12:
    print(data)
```

thetazero commented on rexdb.py:
```diff
@@ -245,5 +458,52 @@ def col(self, i):
+            raise ValueError("time is before database init time")
+        try:
+            with open(filepath, "rb") as fd:
+                for _ in range(self._file_manager.lines_per_file):
```
We can probably make a read file method

thetazero replied:
I think 
```python
for _ in range(self._file_manager.lines_per_file):
                    raw_data = fd.read(self._packer.line_size)
                    data = self._packer.unpack(raw_data)
                    if (data[0] == tfloat):
                        return data
```
is duplicated

tomyyyD replied:
the other part that uses similar logic does not immediately return rather it appends to a list, and the comparisons are slightly different so I don't think i can abstract this to a method.

thetazero replied:
My bad

thetazero commented on rexdb.py:
```diff
@@ -174,63 +217,233 @@ def write_file(self, bytes_data: bytes) -> None:
+        writes a struct of int (file number), float (start time),
+        float (end time) to the map
+        """
+        # print(self.folder_start_time, t)
```
prints should be removed

thetazero reviewed (changes requested):
This is great, a few small tweaks then I think its ready

tomyyyD committed:
Merge pull request #8 from PyCubed-Mini/time_faking

Time faking to speed up unit tests

tomyyyD committed:
cleaning up the filemanager location functions

avoid doing arithmetic in the loops and instead use a while loop and read 12 bytes at a time

tomyyyD committed:
repo cleanup

moving file manager and dense packer into their own files to declutter RexDB

tomyyyD committed:
debugging the cleanup

making things import correctly

tomyyyD committed:
code cleanup

tomyyyD committed:
Cleanup the file names

files do not have folder number in them

tomyyyD committed:
Revert "cleaning up the filemanager location functions"

This reverts commit 329072a49417da48986f2c0e83d833a3122bd0e6.

tomyyyD committed:
Merge branch 'maps' of https://github.com/PyCubed-Mini/RexDB into maps

tomyyyD committed:
reverting to not split into multiple files

I will split into multiple files in a different PR

tomyyyD committed:
cleanup

thetazero reviewed (approved):
Looks good
